File: packages/core/src/storage/S3Provider.js
import { S3Client, PutObjectCommand, ListObjectsV2Command, DeleteObjectCommand } from '@aws-sdk/client-s3';
import StorageProvider from './StorageProvider.js';

export default class S3Provider extends StorageProvider {
  constructor(options) {
    super(options);
    
    if (!options.location) {
      throw new Error('S3Provider requires a "location" (bucket name) in the configuration.');
    }
    
    this.bucket = options.location;
    this.s3Client = new S3Client({
      // The region can be specified in the config or through standard AWS env vars
      region: options.region || process.env.AWS_REGION,
    });
  }

  async save(fileName, data) {
    const params = {
      Bucket: this.bucket,
      Key: fileName,
      Body: data,
      // ACL: 'public-read', // Uncomment if you want objects to be publicly accessible
    };

    try {
      await this.s3Client.send(new PutObjectCommand(params));
      // Return an S3 URI for the object
      return `s3://${this.bucket}/${fileName}`;
    } catch (error) {
      console.error(`Failed to upload ${fileName} to S3:`, error);
      throw error;
    }
  }

  async cleanup() {
    if (!this.options.ttl) {
      console.log('Cleanup skipped: TTL not configured for S3Provider.');
      return;
    }

    const now = Date.now();
    const ttlMillis = this.options.ttl * 1000;

    try {
      const listResponse = await this.s3Client.send(new ListObjectsV2Command({ Bucket: this.bucket }));
      if (!listResponse.Contents) {
        return;
      }

      for (const object of listResponse.Contents) {
        if (now - object.LastModified.getTime() > ttlMillis) {
          console.log(`Cleaning up expired S3 object: ${object.Key}`);
          await this.s3Client.send(new DeleteObjectCommand({ Bucket: this.bucket, Key: object.Key }));
        }
      }
    } catch (error) {
      console.error('An error occurred during S3 cleanup:', error);
    }
  }
}

File: python/pagesnap/storage/s3_provider.py
import boto3
from .storage_provider import StorageProvider

class S3Provider(StorageProvider):
    def __init__(self, options: dict):
        super().__init__(options)
        if "location" not in options:
            raise ValueError("S3Provider requires a 'location' (bucket name) in the configuration.")
        
        self.bucket = options["location"]
        self.s3_client = boto3.client(
            's3',
            region_name=options.get("region") # boto3 handles env vars automatically
        )

    async def save(self, file_name: str, data: bytes) -> str:
        try:
            self.s3_client.put_object(Bucket=self.bucket, Key=file_name, Body=data)
            return f"s3://{self.bucket}/{file_name}"
        except Exception as e:
            print(f"Failed to upload {file_name} to S3: {e}")
            raise
